### PR TITLE
feat(compose): Add Jetpack Compose format

### PR DESCRIPTION
**Pull request by bherbst**
*Issue #*: #478

*Description of changes:*
Add support for Jetpack Compose output via a new `compose/object` format and associated transforms.

**Note**: While Google is developing Compose with Android as the primary target, there is significant interest in supporting other platforms as well. JetBrains is already working with Google to support [desktop applications](https://www.jetbrains.com/lp/compose/), there is some experimental [web support](https://github.com/JetBrains/compose-for-web-demos), and there are people who have been exploring iOS as well. 

As a result, I have explicitly given the format and transforms names that do **not** reference Android. The only Android-specific bits in here right now are the Kotlin imports for the Compose classes and functions which are under  `androidx.compose.ui`. Google has indicated they are working on better [non-Android packaging](https://twitter.com/JimSproch/status/1380276540924039168) so I expect this to evolve.

### `compose/object` format
The `compose/object` format supports outputting all token types to a Kotlin `object`.

```kotlin

package com.example.tokens;

import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.*

object StyleDictionary {
  val colorBackgroundDanger = Color(0xffffeae9)
  val sizeFontLarge = 24.00.sp
  val sizePaddingLarge = 16.00.dp
}
```

### Transforms
The new `compose` transform group includes four new transforms:

 * `color/composeColor` (e.g. `val backgroundRed = Color(0xFFAA0000)`)
 * `size/compose/em` (e.g. `val spacingLarge = 28.em`)
 * `size/compose/remToSp` (e.g. `val spacingLarge = 26.sp`)
 * `size/compose/remToDp` (e.g. `val spacingLarge = 16.00.dp`)
 
 ### Future work:
 There are a few things I'd like to improve in the future and would love feedback on:
 
 * I would like to only add the imports we actually need in the format. Ideally the transforms would be able to tell the format which imports it needs (e.g. the `size/compose/remToSp` transform is the only one that would need the `androidx.compose.ui.unit.sp` import). This type of communication between the format and the transform isn't really supported in Style Dictionary right now.
 * Support for `themeable` property and/or Kotlin classes. It would be very easy to add a Kotlin class format that would support overriding a property, but I only think that would be appropriate for `themeable` properties This becomes somewhat tricky if the input contains a mix of themeable and non-themeable properties.
 * Dark mode - I'm not sure if Style Dictionary intends to support dark mode natively, or if this is something that will be left to examples that leverage custom transforms & formats. I have an initial pass at such an implementation here: https://github.com/amzn/style-dictionary/issues/478#issuecomment-812085193

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: lib/common/templates/compose/object.kt.template [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.unit.*
+
+object <%= file.className %> {
+  <%= allProperties.map(function(prop) {
+    let comment = "";
+    if (prop.comment) {
+      comment = `/** ${prop.comment} */\n  `
+    }
+    return `${comment}val ${prop.name} = ${prop.value}`;
```
Comment (dbanksdesign) [inline lib/common/templates/compose/object.kt.template:30]: Here is where you would use the `formatProperty(prop)` function you built in the format function. This is how you could support outputReferences in this format.

**Review by dbanksdesign — COMMENTED** [review]
This looks great, thank you so much for all your work on this. I just had a minor comment on supporting outputReferences in the format.

File: lib/common/formats.js [hunk 2]
```diff
@@ -549,6 +549,51 @@ module.exports = {
...
+   * }
+   * ```
+   */
+  'compose/object': function({dictionary, options, file}) {
+    const template = _.template(
+      fs.readFileSync(__dirname + '/templates/compose/object.kt.template')
+    );
+    let allProperties;
+    const { outputReferences } = options;
+    const formatProperty = createPropertyFormatter({
```
[the remark this answers is not in the document]
Comment (bherbst) [reply lib/common/formats.js:580]: Oops, you are definitely correct there- I had switched from using the property formatter to doing it all in the template so I could output the comment _before_ the variable (the formatter's internals currently put it _after_ the variable). 

So I did have references working (and intended that to be the case), but forgot to move resolution. I'll revisit that, good catch!

[the remark this answers is not in the document]
Comment (bherbst) [reply lib/common/formats.js:580]: Speaking of which- the current property formatter does some switching on the different CSS outputs, but I was a little hesitant to add kotlin-specific handling as well. For something like moving the comments to before a variable vs. after do you have a preference of updating the formatter or handling it in the template?

File: docs/transforms.md [hunk 3]
```diff
@@ -284,6 +284,20 @@ Transforms the value into an 8-digit hex string for Android because they put the
 ```
 
 
+* * *
+
+### color/composeColor
```
[the remark this answers is not in the document]
Comment (bherbst) [reply docs/transforms.md:289]: Correct!

**Comment by bherbst** [comment]
I have updated the compose format to use the formatter so reference are now getting output again. I ended up taking a middle path with the property formatter- I added support for a "none" comment style that suppresses the comment in the formatter.

Adding a "kotlin" or "javadoc" comment style to the existing set of "short" and "long" didn't feel right.

**Comment by dbanksdesign** [comment]
@bherbst I love the middle ground you took. I think it makes sense to me and balances the right level of flexibility in the format and formatHelper code. Great work!

**Review by chazzmoney — APPROVED** [review]
LGTM! :shipit:

**Pull request merged**